lightning/src/onion_message/messenger.rs

```diff
@@ -659,78 +639,108 @@ where
+			&self.network_graph,
 			recipient,
 			context,
 			peers.into_iter(),
-			entropy_source,
+			&self.entropy_source,
 			secp_ctx,
 			false,
 		)
 	}
+}
 
-	pub(crate) fn create_compact_blinded_paths<T: secp256k1::Signing + secp256k1::Verification>(
-		network_graph: &G, recipient: PublicKey, context: MessageContext,
-		peers: Vec<MessageForwardNode>, entropy_source: &ES, secp_ctx: &Secp256k1<T>,
-	) -> Result<Vec<BlindedMessagePath>, ()> {
-		Self::create_blinded_paths_from_iter(
-			network_graph,
-			recipient,
-			context,
-			peers.into_iter(),
-			entropy_source,
-			secp_ctx,
-			true,
-		)
+/// A [`MessageRouter`] that can only route to a directly connected [`Destination`],
+/// and tries to create compact blinded path for that purpose.
+///
+/// # Privacy
+///
+/// Creating [`BlindedMessagePath`]s may affect privacy since, if a suitable path cannot be found,
+/// it will create a one-hop path using the recipient as the introduction node if it is a announced
+/// node. Otherwise, there is no way to find a path to the introduction node in order to send a
+/// message, and thus an `Err` is returned.
+pub struct CompactMessageRouter<G: Deref<Target = NetworkGraph<L>>, L: Deref, ES: Deref>
+where
+	L::Target: Logger,
+	ES::Target: EntropySource,
+{
+	network_graph: G,
+	entropy_source: ES,
+}
+
+impl<G: Deref<Target = NetworkGraph<L>>, L: Deref, ES: Deref> CompactMessageRouter<G, L, ES>
+where
+	L::Target: Logger,
+	ES::Target: EntropySource,
+{
+	/// Creates a [`DefaultMessageRouter`] using the given [`NetworkGraph`].
+	pub fn new(network_graph: G, entropy_source: ES) -> Self {
+		Self { network_graph, entropy_source }
 	}
 }
 
 impl<G: Deref<Target = NetworkGraph<L>>, L: Deref, ES: Deref> MessageRouter
-	for DefaultMessageRouter<G, L, ES>
+	for CompactMessageRouter<G, L, ES>
 where
 	L::Target: Logger,
 	ES::Target: EntropySource,
 {
 	fn find_path(
 		&self, sender: PublicKey, peers: Vec<PublicKey>, destination: Destination,
 	) -> Result<OnionMessagePath, ()> {
-		Self::find_path(&self.network_graph, sender, peers, destination)
+		DefaultMessageRouter::<G, L, ES>::find_path(&self.network_graph, sender, peers, destination)
 	}
 
 	fn create_blinded_paths<T: secp256k1::Signing + secp256k1::Verification>(
-		&self, recipient: PublicKey, context: MessageContext, peers: Vec<PublicKey>,
+		&self, recipient: PublicKey, context: MessageContext, peers: Vec<MessageForwardNode>,
 		secp_ctx: &Secp256k1<T>,
 	) -> Result<Vec<BlindedMessagePath>, ()> {
-		Self::create_blinded_paths(
+		DefaultMessageRouter::create_blinded_paths_from_iter(
 			&self.network_graph,
 			recipient,
 			context,
-			peers,
+			peers.into_iter(),
 			&self.entropy_source,
 			secp_ctx,
+			true,
 		)
 	}
+}
 
-	fn create_compact_blinded_paths<T: secp256k1::Signing + secp256k1::Verification>(
-		&self, recipient: PublicKey, context: MessageContext, peers: Vec<MessageForwardNode>,
-		secp_ctx: &Secp256k1<T>,
+/// A special [`MessageRouter`] that can doesn't route.
```

Comment:
We should note that this cannot be used to parameterize `ChannelManager` as it wouldn't be able to create reply paths.

Reply:
Done! [pr3246.19](https://github.com/shaavan/rust-lightning/commits/pr3246.19)
Thanks for the pointer!